posthog/hogql/database.py

```diff
@@ -74,25 +115,98 @@ def clickhouse_table(self):
         return "person"
 
 
+def join_with_persons_table(from_table: str, to_table: str, requested_fields: List[str]):
+    from posthog.hogql import ast
+
+    if not requested_fields:
+        raise ValueError("No fields requested from persons table. Why are we joining it?")
+
+    # contains the list of fields we will select from this table
+    fields_to_select: List[ast.Expr] = []
+
+    argmax_version: Callable[[ast.Expr], ast.Expr] = lambda field: ast.Call(
+        name="argMax", args=[field, ast.Field(chain=["version"])]
+    )
+    for field in requested_fields:
+        if field != "id":
+            fields_to_select.append(ast.Alias(alias=field, expr=argmax_version(ast.Field(chain=[field]))))
+
+    id = ast.Field(chain=["id"])
+
+    return ast.JoinExpr(
+        join_type="INNER JOIN",
+        table=ast.SelectQuery(
+            select=fields_to_select + [id],
+            select_from=ast.JoinExpr(table=ast.Field(chain=["persons"])),
+            group_by=[id],
+            having=ast.CompareOperation(
+                op=ast.CompareOperationType.Eq,
+                left=argmax_version(ast.Field(chain=["is_deleted"])),
+                right=ast.Constant(value=0),
+            ),
+        ),
+        alias=to_table,
+        constraint=ast.CompareOperation(
+            op=ast.CompareOperationType.Eq,
+            left=ast.Field(chain=[from_table, "person_id"]),
+            right=ast.Field(chain=[to_table, "id"]),
+        ),
+    )
+
+
 class PersonDistinctIdTable(Table):
     team_id: IntegerDatabaseField = IntegerDatabaseField(name="team_id")
     distinct_id: StringDatabaseField = StringDatabaseField(name="distinct_id")
     person_id: StringDatabaseField = StringDatabaseField(name="person_id")
     is_deleted: BooleanDatabaseField = BooleanDatabaseField(name="is_deleted")
     version: IntegerDatabaseField = IntegerDatabaseField(name="version")
 
+    person: LazyTable = LazyTable(from_field="person_id", table=PersonsTable(), join_function=join_with_persons_table)
+
+    def avoid_asterisk_fields(self):
+        return ["is_deleted", "version"]
+
     def clickhouse_table(self):
         return "person_distinct_id2"
 
 
-class EventsPersonSubTable(Table):
-    id: StringDatabaseField = StringDatabaseField(name="person_id")
-    created_at: DateTimeDatabaseField = DateTimeDatabaseField(name="person_created_at")
-    properties: StringJSONDatabaseField = StringJSONDatabaseField(name="person_properties")
-
-    def clickhouse_table(self):
-        # This is a bit of a hack to make sure person.properties.x works
-        return "events"
+def join_with_max_person_distinct_id_table(from_table: str, to_table: str, requested_fields: List[str]):
```

Comment:
it's maybe too much refactoring for now but this method and `join_with_persons_table` are almost identical and feel like they'd change together... but they're also almost next to each other in a single file so 🤷‍♀️

Reply:
This will change even more when I add support for querying ["is person in cohort"](https://github.com/PostHog/posthog/compare/events-table-based-on-hogql...cohort-filter?expand=1), which needs a similar join. So I'd punt on refactoring this file much further for now. The repetitive patterns and absstractions will emerge as soon as we add a few more joins.